src/codeeditor/editor.ts

```diff
@@ -0,0 +1,588 @@
+
+    /**
+     * The currently selected code.
+     *
+     * @returns A read-only copy of the text selections.
+     */
+    readonly selections: ObservableVector<ITextSelection>;
+
+    /**
+     * Returns the content for the given line number.
+     */
+    getLine(line: number): string;
+
+    /**
+     * Find an offset for the given position.
+     */
+    getOffsetAt(position: IPosition): number;
+
+    /**
+     * Find a position fot the given offset.
+     */
+    getPositionAt(offset: number): IPosition;
+
+    /**
+     * Undo one edit (if any undo events are stored).
+     */
+    undo(): void;
+
+    /**
+     * Redo one undone edit.
+     */
+    redo(): void;
+
+    /**
+     * Clear the undo history.
+     */
+    clearHistory(): void;
+  }
+
+  /**
+   * A keydown handler type.
+   */
+  export
+  type KeydownHandler = (instance: IEditor, event: KeyboardEvent) => boolean;
+
+  /**
+   * A widget that provides a code editor.
+   */
+  export
+  interface IEditor extends Widget {
+    /**
+     * Whether line numbers should be displayed. Defaults to false.
+     */
+    lineNumbers: boolean;
+
+    /**
+     * Set to false for horizontal scrolling. Defaults to true.
+     */
+    wordWrap: boolean;
+
+    /**
+     * Whether the editor is read-only.  Defaults to false.
+     */
+    readOnly: boolean;
+
+    /**
+     * The model used by the editor.
+     */
+    readonly model: IModel;
+
+    /**
+     * The height of a line in the editor in pixels.
+     */
+    readonly lineHeight: number;
+
+    /**
+     * The widget of a character in the editor in pixels.
+     */
+    readonly charWidth: number;
+
+    /**
+     * Handle keydown events for the editor.
+     *
+     * #### Notes
+     * Return `true` to prevent the default handling of the event by the
+     * editor.
+     */
+    onKeyDown: KeydownHandler;
+
+    /**
+     * Brings browser focus to this editor text.
+     */
+    focus(): void;
+
+    /**
+     * Test whether the editor has keyboard focus.
+     */
+    hasFocus(): boolean;
+
+    /**
+     * Set the size of the editor in pixels.
+     */
+    setSize(width: number, height: number): void;
+
+    /**
+     * Scroll the given cursor position into view.
+     */
+    scrollIntoView(pos: IPosition, margin?: number): void;
+
+    /**
+     * Get the window coordinates given a cursor position.
+     */
+    getCoords(position: IPosition): ICoords;
+  }
+
+  /**
+   * The options used to initialize an editor.
+   */
+  export
+  class IOptions {
+    /**
+     * Whether line numbers should be displayed. Defaults to false.
+     */
+    lineNumbers?: boolean;
+
+    /**
+     * Set to false for horizontal scrolling. Defaults to true.
+     */
+    wordWrap?: boolean;
+
+    /**
+     * Whether the editor is read-only.  Defaults to false.
+     */
+    readOnly?: boolean;
+  }
+
+  /**
+   * The default implementation of the code editor model.
+   */
+  export
+  class Model implements IModel {
```

Comment:
Is the idea here to have a specialization of `Model` for CodeMirror in order to keep it in-sync with `CodeMirror.Doc`?